File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static int[][] board;
	static int white = 0;
	static int blue = 0;

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		board = new int[n][n];

		for (int i = 0; i < n; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine(), " ");
			for (int j = 0; j < n; j++) {
				board[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		quadrant(0, 0, n);

		System.out.println(white);
		System.out.println(blue);
	}

	private static void quadrant(int row, int col, int size) {

		if (isSame(row, col, size)) {
			if (board[row][col] == 0) {
				white++;
			} else {
				blue++;
			}
		} else {
			int newSize = size / 2;

			quadrant(row, col, newSize);
			quadrant(row, col + newSize, newSize);
			quadrant(row + newSize, col, newSize);
			quadrant(row + newSize, col + newSize, newSize);
		}
	}

	private static boolean isSame(int row, int col, int size) {

		int color = board[row][col];

		for (int i = row; i < row + size; i++) {
			for (int j = col; j < col + size; j++) {
				if (board[i][j] != color) {
					return false;
				}
			}
		}

		return true;
	}
}

File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.py
import sys

n = int(sys.stdin.readline().rstrip())
white = 0
blue = 0
board = [list(map(int, sys.stdin.readline().rstrip().split())) for _ in range(n)]


def is_same(row, col, size):
    color = board[row][col]
    for i in range(row, row + size):
        for j in range(col, col + size):
            if board[i][j] != color:
                return False
    return True


def quanrant(row, col, size):
    global white, blue
    if is_same(row, col, size):
        if board[row][col] == 0:
            white += 1
        else:
            blue += 1
    else:
        newSize = size // 2

        quanrant(row, col, newSize)
        quanrant(row, col + newSize, newSize)
        quanrant(row + newSize, col, newSize)
        quanrant(row + newSize, col + newSize, newSize)


quanrant(0, 0, n)

print(white)
print(blue)
